env/env.cc

```diff
@@ -29,13 +30,43 @@
 
 namespace ROCKSDB_NAMESPACE {
 namespace {
+#ifndef ROCKSDB_LITE
+static int RegisterBuiltinEnvs(ObjectLibrary& library,
+                               const std::string& /*arg*/) {
+  library.Register<Env>(MockEnv::kClassName(), [](const std::string& /*uri*/,
+                                                  std::unique_ptr<Env>* guard,
+                                                  std::string* /* errmsg */) {
+    guard->reset(MockEnv::Create(Env::Default()));
+    return guard->get();
+  });
+  library.Register<Env>(
+      CompositeEnvWrapper::kClassName(),
+      [](const std::string& /*uri*/, std::unique_ptr<Env>* guard,
+         std::string* /* errmsg */) {
+        guard->reset(new CompositeEnvWrapper(Env::Default()));
+        return guard->get();
+      });
+  size_t num_types;
+  return static_cast<int>(library.GetFactoryCount(&num_types));
+}
+#endif  // ROCKSDB_LITE
+
+static void RegisterSystemEnvs() {
+#ifndef ROCKSDB_LITE
+  static std::once_flag loaded;
+  std::call_once(loaded, [&]() {
+    RegisterBuiltinEnvs(*(ObjectLibrary::Default().get()), "");
+  });
+#endif  // ROCKSDB_LITE
+}
+
 class LegacySystemClock : public SystemClock {
  private:
   Env* env_;
 
  public:
   explicit LegacySystemClock(Env* env) : env_(env) {}
-  const char* Name() const override { return "Legacy System Clock"; }
```

Comment:
Why we remove the 2 names and just give ""?

Reply:
I do not want the "Legacy" name to appear in a serialization of an Env, nor do I want to register it with the infrastructure.  

For example, say there is some class "CustomEnv" that does not have its own FileSystem or SystemClock.  If I serialize it on its own, it will be "id=CustomEnv". However, if I put the CustomEnv inside a Composite (say replace the FileSystem but not the SystemClock), I would get "id=Composite; FileSystem=MyFileSystem; target=CustomEnv; clock=LegacySystemClock".  I do not want/need to use this legacy system clock at all.

If you prefer, I can squash the serialization another way and leave the name.

Reply:
Since we will have an major release, it is ok to make these changes. Just make sure you indicate it in the HISTORY.md